commands/src/main/java/discord4j/commands/CommandException.java

```diff
@@ -0,0 +1,40 @@
+/*
+ *  This file is part of Discord4J.
+ *
+ * Discord4J is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Discord4J is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Discord4J.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package discord4j.commands;
+
+import javax.annotation.Nullable;
+import java.util.Optional;
+
+/**
+ * An abstract class for propagating non-fatal, user-facing, command-caused errors.
+ */
+public class CommandException extends RuntimeException {
```

Comment:
Maybe provide all 4 constructor variants.

Reply:
The purpose of this is to be an error signal though, not an actual exception

Reply:
I need a demonstration of its intended purpose then because wouldn't reactor treat it like an error signal?

Reply:
Check the test directory, the bootstrapper listens for these specific error signals and then invokes the error handler, because the point of this specific exception is to allow for unified error reporting by a bot, regardless of who implemented a specific command

Reply:
Two questions then:
1) Why not just handle all exceptions this way, i.e. why does doOnError have to be specific to our exception and not just handle all exceptions?
2) Then why not allow to carry other exceptions like, say, a database failure. In the example it shows a `switchIfEmpty` but what if my logic threw a `SQLException`. I would have to log the exception *then* throw this exception, when I can just simply wrap it.

Reply:
1. The end user should use reactor apis for handling generic exceptions as the resultant flux is exposed, therefore allowing users to do this. The specific problem this is trying to solve is that if a bot incorporates multiple external command impls, they may be inconsistent in *user-facing* error reporting (for example, one may DM the user, another may send a reply, etc), but this allows for more consistent error reporting.
2. I think its perfectly reasonable for a user to wrap exceptions into a new CommandException object as since its purpose is end user-facing messages, a raw stacktrace is generally not desireable.

Reply:
1. I still don't see how the exception specifically accomplishes this. If you want unified error handling then set the error handler and handle whatever exceptions propagate down.

2. That's what I'm saying, you can't wrap exceptions, there's no available constructor here. The error handler should be the one to handle how the error is outputted to the user-facing system, not by this Exception restricting what you can or can not send.

Reply:
1. Correction, this is unified error handling for user-facing errors specifically. So all commands, regardless of dev, will report errors like missing permissions or something in a consistent way. 
2. But again that misses the point of the exception, it's just a signal to tell the command handler to message in some way that a problem has occurred that is non-fatal.